File: 2022/day15/d15.php
<?php

declare(strict_types=1);

readonly class Sensor
{
    private int $manathanDistanceFromBeacon;
    public function __construct(
        public int $x,
        public int $y,
        public int $beaconX,
        public int $beaconY
    ) {
        $this->manathanDistanceFromBeacon = abs($this->x - $this->beaconX) + abs($this->y - $this->beaconY);
    }

    public function getManathanDistanceFromBeacon(): int
    {
        return $this->manathanDistanceFromBeacon;
    }

    public function getManathanDistanceFrom(int $x, int $y): int
    {
        return abs($this->x - $x) + abs($this->y - $y);
    }

    public function getSurroundingPoints(int $manathanDistance, int $lowerBound = 0, int $higherBound = 4000000): iterable
    {
        $minY = max($lowerBound, $this->y - $manathanDistance);
        $maxY = min($higherBound, $this->y + $manathanDistance);
        for ($y = $minY; $y <= $maxY; ++$y) {
            $hd = abs(abs($this->y - $y) - $manathanDistance);
            for ($x = $this->x - $hd; $x <= $this->x + $hd; $x+=(2*$hd)) {
                if ($x >= $lowerBound && $x <= $higherBound) {
                    yield [$x, $y];
                }
                if (0 === $hd) {
                    break;
                }
            }
        }
    }
}

readonly class Map
{
    public function __construct(
        private array $sensors
    ) {
    }

    public function countPositionsNotHavingBeaconsOnLine(int $y): int
    {
        $ranges = [];
        foreach ($this->sensors as $sensor) {
            $md = $sensor->getManathanDistanceFromBeacon();

            if ($y >= $sensor->y - $md && $y <= $sensor->y + $md) {
                $hd = abs(abs($sensor->y - $y) - $md);
                if ($y === $sensor->beaconY && $sensor->beaconX >= $sensor->x - $hd && $sensor->beaconX <= $sensor->x + $hd) {
                    $ranges[] = [$sensor->x - $hd, $sensor->beaconX - 1];
                    $ranges[] = [$sensor->beaconX + 1, $sensor->x + $hd];
                } else {
                    $ranges[] = [$sensor->x - $hd, $sensor->x + $hd];
                }
            }
        }

        usort($ranges, fn($a, $b) => $a[0] <=> $b[0]);
        $beaconsFreePos = 0;
        $previous = array_shift($ranges);
        while ($range = array_shift($ranges)) {
            if ($range[0] <= $previous[1] && $range[1] > $previous[1]) {
                $previous[1] = $range[1];
            } elseif ($range[0] > $previous[1]) {
                $beaconsFreePos += $previous[1] - $previous[0] + 1;
                $previous = $range;
            }
        }
        $beaconsFreePos += $previous[1] - $previous[0] + 1;

        return $beaconsFreePos;
    }

    private function canHaveBeacon(int $x, int $y): bool
    {
        foreach ($this->sensors as $sensor) {
            if ($sensor->getManathanDistanceFromBeacon() >= $sensor->getManathanDistanceFrom($x, $y)) {
                return false;
            }
        }

        return true;
    }

    public function getTuningFrequency(int $lowerBound = 0, int $higherBound = 4000000): int
    {
        foreach ($this->sensors as $sensor) {
            $md = $sensor->getManathanDistanceFromBeacon() + 1;

            foreach ($sensor->getSurroundingPoints($md, $lowerBound, $higherBound) as [$x, $y]) {
                if ($this->canHaveBeacon($x, $y)) {
                    return $x * 4000000 + $y;
                }
            }
        }

        return -1;
    }
}

$input = trim(file_get_contents('php://stdin'));
$input = explode(PHP_EOL, $input);

$sensors = [];
foreach ($input as $line) {
    sscanf($line, 'Sensor at x=%d, y=%d: closest beacon is at x=%d, y=%d', $x, $y, $bx, $by);
    $sensors[] = new Sensor($x, $y, $bx, $by);
}

$map = new Map($sensors);
printf(
    "Result 1: %d\nResult 2: %d\n",
    $map->countPositionsNotHavingBeaconsOnLine((int) $_SERVER['argv'][1] ?? 10),
    $map->getTuningFrequency(higherBound: (int) $_SERVER['argv'][2] ?? 20)
);

File: 2022/day15/d15.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

using namespace std;

class Sensor {
public:
  Sensor(int x, int y, int beaconX, int beaconY) {
    this->x = x;
    this->y = y;
    this->beaconX = beaconX;
    this->beaconY = beaconY;
	this->distance = abs(beaconX - x) + abs(beaconY - y);
  }
  int x;
  int y;
  int beaconY;
  int beaconX;
  int distance;
};

class Map {
private:
  vector<Sensor> sensors;

public:
  Map(vector<Sensor> sensors) { this->sensors = sensors; }

  int countPositionsNotHavingBeaconsOnLine(int y) {
    int freePositions = 0;
    vector<pair<int, int>> ranges;

    for (Sensor sensor : sensors) {
      if (y >= sensor.y - sensor.distance && y <= sensor.y + sensor.distance) {
        int hd = abs(abs(sensor.y - y) - sensor.distance);
        if (sensor.beaconY == y && sensor.beaconX >= sensor.x - hd &&
            sensor.beaconX <= sensor.x + hd) {
          ranges.push_back(pair<int, int>(sensor.x - hd, sensor.beaconX - 1));
          ranges.push_back(pair<int, int>(sensor.beaconX + 1, sensor.x + hd));
        } else {
          ranges.push_back(pair<int, int>(sensor.x - hd, sensor.x + hd));
        }
      }
    }

    sort(ranges.begin(), ranges.end());

    if (ranges.empty()) {
      return 0;
    }

    pair<int, int> previous = ranges[0];
    ranges.erase(ranges.begin());
    for (pair<int, int> range : ranges) {
      if (range.first <= previous.second && range.second > previous.second) {
        previous = {previous.first, range.second};
      } else if (range.first > previous.second) {
        freePositions += previous.second - previous.first + 1;
        previous = range;
      }
    }

    freePositions += previous.second - previous.first + 1;

    return freePositions;
  }

  bool canHaveBeacon(int x, int y) {
    for (Sensor sensor : sensors) {
      if (sensor.distance >= abs(x - sensor.x) + abs(y - sensor.y)) {
        return false;
      }
    }
    return true;
  }

  long long getTuningFrequency(int lowerBound, int higherBound) {
    for (Sensor sensor : sensors) {
      int md = sensor.distance + 1;
      int minY = max(lowerBound, sensor.y - md);
      int maxY = min(higherBound, sensor.y + md);
      for (int y = minY; y <= maxY; y++) {
        int hd = abs(abs(sensor.y - y) - md);
        for (int x = sensor.x - hd; x <= sensor.x + hd; x += 2 * hd) {
          if (x >= lowerBound && x <= higherBound && canHaveBeacon(x, y)) {
            return (long long)x * 4000000 + y;
          } else if (0 == hd) {
            break;
          }
        }
      }
    }
    return -1;
  }
};

int main(int argc, char **argv) {
  string line;
  vector<Sensor> sensors;

  while (getline(cin, line)) {
    int x, y, beaconX, beaconY;
    sscanf(line.c_str(), "Sensor at x=%d, y=%d: closest beacon is at x=%d, y=%d", &x, &y, &beaconX, &beaconY);
    sensors.push_back(Sensor(x, y, beaconX, beaconY));
  }

  Map map(sensors);
  cout << "Result 1: " << map.countPositionsNotHavingBeaconsOnLine(argc > 1 ? stoi(argv[1]) : 2000000) << endl;
  cout << "Result 2: " << map.getTuningFrequency(0, argc > 2 ? stoi(argv[2]) : 4000000) << endl;
}
